proof/refine/ARM/Reply_R.thy

```diff
@@ -634,4 +618,453 @@ lemma replyRemoveTCB_invs':
+        apply (clarsimp simp: replies_with_sc_def image_def)
+        apply (drule_tac x=scptr in spec)
+        apply (clarsimp simp: sc_at_pred_n_def obj_at_def)
+       apply (erule (1) sc_replies_relation_prevs_list)
+       apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKO_sc)
+      apply (erule (1) sc_replies_relation_scReply[symmetric])
+      apply (clarsimp simp: obj_at'_real_def  ko_wp_at'_def projectKO_sc)
+     apply wpsimp+
+  done
+
+lemma setObject_reply_pde_mappings'[wp]:
+  "setObject ptr (val :: reply) \<lbrace>valid_pde_mappings'\<rbrace>"
+  by (wp valid_pde_mappings_lift')
+
+crunches bindScReply
+  for pspace_aligned'[wp]: pspace_aligned'
+  and pspace_distinct'[wp]: pspace_distinct'
+  and if_unsafe_then_cap'[wp]: if_unsafe_then_cap'
+  and valid_global_refs'[wp]: valid_global_refs'
+  and valid_arch_state'[wp]: valid_arch_state'
+  and valid_irq_node'[wp]: "\<lambda>s. valid_irq_node' (irq_node' s) s"
+  and irq_node'[wp]: "\<lambda>s. P (irq_node' s)"
+  and valid_irq_handlers'[wp]: valid_irq_handlers'
+  and valid_irq_states'[wp]: valid_irq_states'
+  and valid_machine_state'[wp]: valid_machine_state'
+  and irqs_masked'[wp]: irqs_masked'
+  and valid_release_queue'[wp]: valid_release_queue'
+  and ct_not_inQ[wp]: ct_not_inQ
+  and ct_idle_or_in_cur_domain'[wp]: ct_idle_or_in_cur_domain'
+  and valid_pde_mappings'[wp]: valid_pde_mappings'
+  and pspace_domain_valid[wp]: pspace_domain_valid
+  and ksCurDomain[wp]: "\<lambda>s. P (ksCurDomain s)"
+  and valid_dom_schedule'[wp]: valid_dom_schedule'
+  and cur_tcb'[wp]: cur_tcb'
+  and no_0_obj'[wp]: no_0_obj'
+  and valid_mdb'[wp]: valid_mdb'
+  and tcb_at'[wp]: "tcb_at' t"
+  and cte_wp_at'[wp]: "cte_wp_at' P p"
+  and ctes_of[wp]: "\<lambda>s. P (ctes_of s)"
+  (wp: crunch_wps hoare_vcg_all_lift valid_irq_node_lift simp: crunch_simps valid_mdb'_def)
+
+crunches rescheduleRequired, scheduleTCB
+  for valid_tcb_state'[wp]: "valid_tcb_state' ts"
+  (wp: crunch_wps)
+
+crunches setThreadState
+  for sc_ko_at'[wp]: "\<lambda>s. P (ko_at' (sc :: sched_context) p s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+crunches updateReply
+  for valid_obj'[wp]: "valid_obj' obj"
+
+lemma updateReply_replyTCB_valid_objs'[wp]:
+  "\<lbrace>valid_objs' and case_option \<top> (\<lambda>t. tcb_at' t) p\<rbrace>
+   updateReply rptr (replyTCB_update (\<lambda>_. p))
+   \<lbrace>\<lambda>_. valid_objs'\<rbrace>"
+  by (wpsimp wp: set_reply'.valid_objs' simp: updateReply_def)
+     (auto simp: valid_obj'_def valid_reply'_def obj_at'_def projectKO_eq projectKO_reply
+          split: option.splits)
+
+lemma updateReply_valid_objs':
+  "\<lbrace>valid_objs'
+    and (\<lambda>s. \<forall>ko. ko_at' ko rptr s \<longrightarrow> valid_reply' (f ko) s)\<rbrace>
+   updateReply rptr f
+   \<lbrace>\<lambda>_. valid_objs'\<rbrace>"
+  apply (wpsimp wp: set_reply'.valid_objs' simp: updateReply_def)
+  by (fastforce simp: valid_obj'_def)
+
+lemma replyNext_update_Head_valid_objs'[wp]:
```

Comment:
This one and a few below which are instantiations for specific cases of the above `updateReply_valid_objs’` are also something we could do without, if we use the `xxx_preserved` lemmas. I don’t think it is a great idea to chucking all these in the wp set, either.

Reply:
These cases come up as nuisance I think because `set_sc_valid_objs’` and the friends are in the wp set. We should take them out of the wp set.

Reply:
Happy to remove them if needed, but can you explain a little more about why they're a nuisance? That surprises me because they look almost like weakest preconditions to me.

Reply:
I am just guessing that you wrote these because you had to stop at these points while working on your main proof (like `bindScReply_valid_objs’ below), rather than just resolving the subgoals in place. But it is probably not because of what I said. It is more because of the `ko_at’` in your version of `updateReply_valid_objs’` that require a bit of work (but not much). With `updateReply_valid_objs’_preserved_strong`, these are just instantiations of it with a bit of unfolding, and I don’t think we need to list all these specific lemmas when we have a more abstract one.

Still, I think we still want to remove `set_sc_valid_objs’` and others from the wp set. They unfold `valid_obj’` unnecessarily, when wpsimp can make more progress with `valid_obj’`.

Reply:
Mitch and I agreed that, for these lemmas, leaving them as they are now would be the best option for now (least work required).